system/jlib/jsocket.cpp

```diff
@@ -1098,12 +1173,63 @@ void CSocket::set_linger(int lingertime)
     }
 }
 
-void CSocket::set_keep_alive(bool set)
+void CSocket::setKeepAlive(bool set, int time, int intvl, int probes)
 {
     int on=set?1:0;
-    if (setsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, (char*)&on, sizeof(on)) != 0) {
+    if (setsockopt(sock, SOL_SOCKET, SO_KEEPALIVE, (char*)&on, sizeof(on)) != 0)
+    {
         IWARNLOG("KeepAlive not set");
+        return;
+    }
+
+    if (!on)
+        return;
+
+    int srtn, optval;
+    socklen_t optlen = sizeof(optval);
+
+    optval = time;
+    srtn = setsockopt(sock, IPPROTO_TCP, TCP_KEEPIDLE, &optval, optlen);
+    if (srtn != 0)
+    {
+        IWARNLOG("KeepAlive time not set");
+        return;
+    }
+
+    optval = intvl;
+    srtn = setsockopt(sock, IPPROTO_TCP, TCP_KEEPINTVL, &optval, optlen);
+    if (srtn != 0)
+    {
+        IWARNLOG("KeepAlive probes not set");
+        return;
     }
+
+    optval = probes;
+    srtn = setsockopt(sock, IPPROTO_TCP, TCP_KEEPCNT, &optval, optlen);
+    if (srtn != 0)
+    {
+        IWARNLOG("KeepAlive probes not set");
+        return;
+    }
+}
+
+void CSocket::checkCfgKeepAlive()
+{
+    int time, intvl, probes;
+    if (queryKeepAlive(time, intvl, probes))
+        setKeepAlive(true, time, intvl, probes);
+}
+
+void CSocket::set_keep_alive(bool set)
+{
+    if (set)
+    {
+        int time, intvl, probes;
+        (void)queryKeepAlive(time, intvl, probes);
```

Comment:
The semantics of (void)queryKeepAlive isn't that intuitive here - it fills out reference params with hard-wired defaults if it returns false.

It would be better I think if this method wasn't used here (see comments below).
And queryKeepAlive() didn't fill in the reference params if returned false, because elsewhere where used, the result is only used if queryKeepAlive returns true.

Reply:
If set_keep_alive(true) is called I think we should use the global keepalive time, interval, probes settings, if present.
So I would like to leave the query in to get those.

Reply:
my earlier point, was that the custom settings will be applied anyway, without setting here in set_keep_alive,
because checkCfgKeepAlive is called per socket before the in open() or pre_connect() and the custom idle, interval and probes will be applied there.
But maybe I missing something?

Reply:
ok, I see your point, if already on then set_keep_alive(true) doesn't really need to do anything, or maybe at the minimum do what it always did.